### PR TITLE
add retry policy when deleting tfjob annotations

### PR DESCRIPTION
**Pull request by OyutianO**

### CODE REVIEW

File: pkg/contorller/contoller.go [hunk 1]
```diff
@@ -423,34 +424,63 @@ func (tc *TFExtensionController) deleteQueueUnitInstance(namespace, name string)
...
+		} else {
+			time.Sleep(RetryInterval)
+			tfJob, err := tc.getTFJob(namespace, tfJobName)
+			if err != nil {
+				klog.Errorf("Get tfjob %v/%v failed when retrying, %v", namespace, tfJobName, err.Error())
+				return err
+			}
+			tfJob.SetAnnotations(annotation)
+			_, err = tc.tfjobClient.KubeflowV1().TFJobs(namespace).Update(context.TODO(), tfJob, metav1.UpdateOptions{})
+			if err !=nil && strings.Contains(err.Error(), OptimisticLockErrorMsg) {
```
Comment (denkensk) [inline pkg/contorller/contoller.go:469]:
```suggestion
			if err != nil && strings.Contains(err.Error(), OptimisticLockErrorMsg) {
```

File: pkg/contorller/contoller.go [hunk 2]
```diff
@@ -423,34 +424,63 @@ func (tc *TFExtensionController) deleteQueueUnitInstance(namespace, name string)
...
 	var annotation = map[string]string{}
 	for k, v := range tfJob.Annotations {
 		if k != Suspend {
 			annotation[k] = v
 		}
 	}
 	tfJob.SetAnnotations(annotation)
-
 	// TODO change to patch
 	_, err = tc.tfjobClient.KubeflowV1().TFJobs(namespace).Update(context.TODO(), tfJob, metav1.UpdateOptions{})
```
Comment (denkensk) [inline pkg/contorller/contoller.go:454]: It is better to wrap Pod's update operation in a function

File: pkg/contorller/contoller.go [hunk 3]
```diff
@@ -423,34 +424,63 @@ func (tc *TFExtensionController) deleteQueueUnitInstance(namespace, name string)
...
+			_, err = tc.tfjobClient.KubeflowV1().TFJobs(namespace).Update(context.TODO(), tfJob, metav1.UpdateOptions{})
+			if err !=nil && strings.Contains(err.Error(), OptimisticLockErrorMsg) {
+				klog.Warningf("Update annotations for tfjob %v/%v failed because of optimisticLock ,retried %d times",
+					namespace, tfJobName, retry + 1)
+			}
+			if err !=nil && !strings.Contains(err.Error(), OptimisticLockErrorMsg) {
+				klog.Errorf("Update annotations for tfjob %v/%v failed,%v", namespace, tfJobName, err.Error())
+				return err
+			}
+			if retry > MaxRetries {
```
Comment (denkensk) [inline pkg/contorller/contoller.go:477]: MaxRetries is 15 ？ It maybe so large...

File: pkg/contorller/contoller.go [hunk 4]
```diff
@@ -423,34 +424,63 @@ func (tc *TFExtensionController) deleteQueueUnitInstance(namespace, name string)
...
+				klog.Errorf("update tfjob %v/%v error after retried: %v", namespace, tfJobName, err.Error())
```
Comment (denkensk) [inline pkg/contorller/contoller.go:478]:
```suggestion
				klog.Errorf("After maximum %v retries, it still fails to update tfjob %v/%v : error: %v", MaxRetries, namespace, tfJobName, err.Error())
```

**Review by denkensk — COMMENTED** [review]
Thanks @OyutianO 
Need to go fmt this file before.

File: pkg/contorller/contoller.go (hunk 2 above)
Comment (OyutianO) [reply pkg/contorller/contoller.go]: agree!

File: pkg/contorller/contoller.go (hunk 3 above)
Comment (OyutianO) [reply pkg/contorller/contoller.go]: ok, change to 5

File: pkg/contorller/contoller.go (hunk 2 above)
Comment (OyutianO) [reply pkg/contorller/contoller.go]: agree!

**Comment by denkensk** [comment]
/lgtm
/approve
@OyutianO  Thanks so much.

**Pull request merged**